lib/datasource/go/base.ts

```diff
@@ -0,0 +1,172 @@
+import { PlatformId } from '../../constants';
+import { logger } from '../../logger';
+import * as hostRules from '../../util/host-rules';
+import { Http } from '../../util/http';
+import { regEx } from '../../util/regex';
+import { trimTrailingSlash } from '../../util/url';
+import * as github from '../github-tags';
+import * as gitlab from '../gitlab-tags';
+import { bitbucket } from './common';
+import type { DataSource } from './types';
+
+// TODO: figure out class hierarchy (#10532)
+export class BaseGoDatasource {
+  private static readonly gitlabHttpsRegExp = regEx(
+    /^(?<httpsRegExpUrl>https:\/\/[^/]*gitlab\.[^/]*)\/(?<httpsRegExpName>.+?)[/]?$/
+  );
+  private static readonly gitlabRegExp = regEx(
+    /^(?<regExpUrl>gitlab\.[^/]*)\/(?<regExpPath>.+?)[/]?$/
+  );
+
+  private static readonly id = 'go';
+  private static readonly http = new Http(BaseGoDatasource.id);
+
+  static async getDatasource(goModule: string): Promise<DataSource | null> {
+    if (goModule.startsWith('gopkg.in/')) {
+      const [pkg] = goModule.replace('gopkg.in/', '').split('.');
+      const lookupName = pkg.includes('/') ? pkg : `go-${pkg}/${pkg}`;
+      return {
+        datasource: github.id,
+        lookupName,
+        registryUrl: 'https://github.com',
+      };
+    }
+
+    if (goModule.startsWith('github.com/')) {
+      const split = goModule.split('/');
+      const lookupName = split[1] + '/' + split[2];
+      return {
+        datasource: github.id,
+        lookupName,
+        registryUrl: 'https://github.com',
+      };
+    }
+
+    if (goModule.startsWith('bitbucket.org/')) {
+      const split = goModule.split('/');
+      const lookupName = split[1] + '/' + split[2];
+      return {
+        datasource: bitbucket.id,
+        lookupName,
+        registryUrl: 'https://bitbucket.org',
+      };
+    }
+
+    const goGetResult = await BaseGoDatasource.goGetDatasource(goModule);
+    return goGetResult;
+  }
+
+  private static async goGetDatasource(
+    goModule: string
+  ): Promise<DataSource | null> {
+    const pkgUrl = `https://${goModule}?go-get=1`;
+    const res = (await BaseGoDatasource.http.get(pkgUrl)).body;
+    const sourceMatch = regEx(
+      `<meta\\s+name="go-source"\\s+content="([^\\s]+)\\s+([^\\s]+)`
+    ).exec(res);
+    if (sourceMatch) {
+      const [, prefix, goSourceUrl] = sourceMatch;
+      if (!goModule.startsWith(prefix)) {
+        logger.trace({ goModule }, 'go-source header prefix not match');
+        return null;
+      }
+      logger.debug({ goModule, goSourceUrl }, 'Go lookup source url');
+      if (goSourceUrl?.startsWith('https://github.com/')) {
+        return {
+          datasource: github.id,
+          lookupName: goSourceUrl
+            .replace('https://github.com/', '')
+            .replace(regEx(/\/$/), ''),
+          registryUrl: 'https://github.com',
+        };
+      }
+      const gitlabUrl =
+        BaseGoDatasource.gitlabHttpsRegExp.exec(goSourceUrl)?.groups
```

Comment:
Please create local const for `BaseGoDatasource.gitlabHttpsRegExp.exec(goSourceUrl)`